bodhi/tests/server/consumers/test_pungi_wrapper.py

```diff
@@ -0,0 +1,419 @@
+            phase_instance = phase["class"]()
+            phase_instance.skip.return_value = False
+
+        wrapper = PungiWrapper(compose, variants_conf)
+
+        wrapper.execute_phases()
+
+        for phase in phases:
+            phase_instance = phase["class"]()
+            phase_instance.validate.assert_called_once()
+
+    @mock.patch("pungi.metadata")
+    @mock.patch("bodhi.server.consumers.pungi_wrapper.PungiWrapper.write_repo_metadata")
+    @mock.patch("bodhi.server.consumers.pungi_wrapper.PungiWrapper.create_latest_repo_links")
+    @mock.patch("pungi.phases")
+    def test_execute_phases_exception(self, pungi_phases, *args):
+        init_phase = pungi_phases.InitPhase()
+        init_phase.validate.side_effect = ValueError("Test Error")
+        init_phase.skip.return_value = False
+        init_phase.name = "init phase"
+        compose = mock.Mock()
+        variants_conf = mock.Mock()
+        wrapper = PungiWrapper(compose, variants_conf)
+
+        with pytest.raises(Exception):
+            wrapper.execute_phases()
+
+    @mock.patch("pungi.util")
+    def test_init_compose_dir(self, util, tmpdir):
+        compose = mock.Mock()
+        variants_conf = mock.Mock()
+        logger = mock.Mock()
+        wrapper = PungiWrapper(compose, variants_conf)
+
+        topdir = str(tmpdir.mkdir("mash-dir"))
+        pungi_conf_path = get_pungi_conf(tmpdir)
+        pungi_conf = PungiConfig(str(pungi_conf_path), logger)
+        compose_id = "f27-modular-updates"
+        compose_dir = wrapper.init_compose_dir(topdir, pungi_conf, compose_id)
+
+        assert compose_dir == os.path.join(topdir, compose_id)
+
+    @mock.patch("bodhi.server.consumers.pungi_wrapper.PungiWrapper.load_variants_config")
+    @mock.patch("bodhi.server.consumers.pungi_wrapper.PungiWrapper.execute_phases")
+    def test_compose_repo(self, load_variants_config, execute_phases):
+        compose = mock.Mock()
+        variants_conf = mock.Mock()
+
+        wrapper = PungiWrapper(compose, variants_conf)
+
+        wrapper.compose_repo()
+
+        load_variants_config.assert_called_once_with()
+        execute_phases.assert_called_once_with()
+
+    @mock.patch("shutil.copyfileobj")
+    def test_load_variants_config(self, copyfileobj, tmpdir):
+        logger = mock.Mock()
+        topdir = str(tmpdir.mkdir("mash_dir"))
+        compose_id = "f27-modular-updates"
+        pungi_conf_path = get_pungi_conf(tmpdir)
+        pungi_conf = PungiConfig(str(pungi_conf_path), logger)
+        compose_dir = PungiWrapper.init_compose_dir(topdir, pungi_conf, compose_id)
+
+        compose = get_compose(pungi_conf, compose_dir)
+        updates, builds = get_updates_and_builds()
+        variants_conf = VariantsConfig(updates, builds)
+        topdir = compose.topdir
+
+        wrapper = PungiWrapper(compose, variants_conf)
+        wrapper.load_variants_config()
+
+        assert len(wrapper.compose.all_variants)
+        copyfileobj.assert_called_once()
```

Comment:
Let's assert the arguments to ```copyfileobj```, and also not just that the len of the ```all_variants``` is non-0.

Reply:
Both are file descriptor object.... not really sure how to do this...